File: MJTableViewController/Demo/AutoHeightDemo/AutoHeightWeiboCell.m
//
//  AutoHeightWeiboCell.m
//  MJTableViewController
//
//  Created by 高扬 on 2018/3/28.
//  Copyright © 2018年 高扬. All rights reserved.
//

#import "AutoHeightWeiboCell.h"
#import "WeiboModel.h"
#import "UIImageView+WebCache.h"

#define TOPIC_AREA_HEIGHT 50
#define IMAGE_AREA_HEIGHT 100
#define LEFT_PADDING 60
#define RIGHT_PADDING 10
#define BOTTOM_PADDING 10

@interface AutoHeightWeiboCell()

@property(nonatomic,retain)UIImageView* iconView;

@property(nonatomic,retain)UILabel* nameLabel;

@property(nonatomic,retain)UILabel* titleLabel;

@property(nonatomic,retain)UILabel* contentLabel;

@property(nonatomic,retain)UIImageView* photoView;

@property(nonatomic,retain)UIView* bottomLine;

@end

@implementation AutoHeightWeiboCell

-(UIImageView *)iconView{
    if (!_iconView) {
        _iconView = [[UIImageView alloc]init];
        _iconView.contentMode = UIViewContentModeScaleAspectFill;
        CGFloat const iconWidth = LEFT_PADDING - 20;
        _iconView.size = CGSizeMake(iconWidth, iconWidth);
        _iconView.layer.cornerRadius = iconWidth / 2.;
        _iconView.layer.masksToBounds = YES;
        [self.contentView addSubview:_iconView];
    }
    return _iconView;
}

-(UIImageView *)photoView{
    if (!_photoView) {
        _photoView = [[UIImageView alloc]init];
        _photoView.contentMode = UIViewContentModeScaleAspectFill;
        _photoView.layer.masksToBounds = YES;
        [self.contentView addSubview:_photoView];
    }
    return _photoView;
}

-(UILabel *)titleLabel{
    if (!_titleLabel) {
        _titleLabel = [UICreationUtils createLabel:SIZE_TEXT_SECONDARY color:COLOR_TEXT_SECONDARY];
        [self.contentView addSubview:_titleLabel];
    }
    return _titleLabel;
}

-(UILabel *)nameLabel{
    if (!_nameLabel) {
        _nameLabel = [UICreationUtils createLabel:SIZE_TEXT_PRIMARY color:COLOR_TEXT_PRIMARY];
        [self.contentView addSubview:_nameLabel];
    }
    return _nameLabel;
}

-(UILabel *)contentLabel{
    if (!_contentLabel) {
        _contentLabel = [UICreationUtils createLabel:SIZE_TEXT_SECONDARY color:COLOR_TEXT_PRIMARY];
        _contentLabel.lineBreakMode = NSLineBreakByWordWrapping;
        _contentLabel.numberOfLines = 0;
        [self.contentView addSubview:_contentLabel];
    }
    return _contentLabel;
}

-(UIView *)bottomLine{
    if (!_bottomLine) {
        _bottomLine = [[UIView alloc]init];
        _bottomLine.backgroundColor = COLOR_LINE;
        _bottomLine.height = LINE_WIDTH;
        [self.contentView addSubview:_bottomLine];
    }
    return _bottomLine;
}

//    NSLog(@"getCellHeight被调用！");
-(CGFloat)getCellHeight:(CGFloat)cellWidth{
    WeiboModel* weiboModel = GET_CELL_DATA(WeiboModel.class);//获取Model
    NSString* content = weiboModel.content;//获取动态内容字符串
    CGRect contentSize = [content boundingRectWithSize:CGSizeMake(cellWidth - LEFT_PADDING - RIGHT_PADDING, FLT_MAX)
                             options:NSStringDrawingUsesLineFragmentOrigin
                          attributes:@{NSFontAttributeName:[UIFont systemFontOfSize:SIZE_TEXT_SECONDARY]}
                             context:nil];//计算给定范围内最佳尺寸
    return TOPIC_AREA_HEIGHT + contentSize.size.height + IMAGE_AREA_HEIGHT + BOTTOM_PADDING * 2;//返回计算后的最终高度
}

-(void)showSubviews{
    self.backgroundColor = [UIColor whiteColor];
    
    WeiboModel* weiboModel = GET_CELL_DATA(WeiboModel.class);
    
    [self.iconView sd_setImageWithURL:[NSURL URLWithString:weiboModel.iconName]];
    self.iconView.centerX = LEFT_PADDING / 2.;
    self.iconView.centerY = TOPIC_AREA_HEIGHT / 2;
    
    self.nameLabel.text = weiboModel.name;
    [self.nameLabel sizeToFit];
    self.titleLabel.text = weiboModel.title;
    [self.titleLabel sizeToFit];
    CGFloat const gap = 5;
    CGFloat const baseY = (TOPIC_AREA_HEIGHT - self.nameLabel.height - gap - self.titleLabel.height) / 2;
    self.nameLabel.y = baseY;
    self.titleLabel.y = self.nameLabel.maxY + gap;
    
    self.nameLabel.x = self.titleLabel.x = LEFT_PADDING;
    
    self.contentLabel.text = weiboModel.content;
    self.contentLabel.size = [self.contentLabel sizeThatFits:CGSizeMake(self.contentView.width - LEFT_PADDING - RIGHT_PADDING, FLT_MAX)];
    self.contentLabel.x = LEFT_PADDING;
    self.contentLabel.y = TOPIC_AREA_HEIGHT;
    
    [self.photoView sd_setImageWithURL:[NSURL URLWithString:weiboModel.imageUrl]];
    self.photoView.frame = CGRectMake(LEFT_PADDING, self.contentLabel.maxY + BOTTOM_PADDING, self.contentLabel.width, IMAGE_AREA_HEIGHT);
    
    self.bottomLine.x = 0;
    self.bottomLine.maxY = self.contentView.height;
    self.bottomLine.width = self.contentView.width;
}

@end

File: GYTableViewController4Swift/GYTableViewController4Swift/Demo/AutoHeightDemo/AutoHeightWeiboCell.swift
//
//  AutoHeightWeiboCell.swift
//  GYTableViewController4Swift
//
//  Created by gaoyang on 2019/2/5.
//  Copyright © 2019年 高扬. All rights reserved.
//

import UIKit

class AutoHeightWeiboCell: GYTableViewCell {

    public static let TOPIC_AREA_HEIGHT:CGFloat = 50
    public static let IMAGE_AREA_HEIGHT:CGFloat = 100
    public static let LEFT_PADDING:CGFloat = 60
    public static let RIGHT_PADDING:CGFloat = 10
    public static let BOTTOM_PADDING:CGFloat = 10

    //MARK: 获取动态高度,高度被缓存不会二次计算
    //    NSLog(@"getCellHeight被调用！");
    override func getHeight(_ cellWidth: CGFloat) -> CGFloat {
        let weiboModel:WeiboModel? = getData() as? WeiboModel//获取Model
        if weiboModel == nil {
            return 0
        }
        
        let content:String  = weiboModel!.content //获取动态内容字符串
        
        let contentSize:CGRect = content.boundingRect(with: CGSize.init(width: cellWidth - AutoHeightWeiboCell.LEFT_PADDING - AutoHeightWeiboCell.RIGHT_PADDING, height: CGFloat(Float.greatestFiniteMagnitude)), options: NSStringDrawingOptions.usesLineFragmentOrigin, attributes: [NSAttributedString.Key.font:UIFont.systemFont(ofSize: TVStyle.sizeTextSecondary)], context: nil)
        //计算给定范围内最佳尺寸
        return AutoHeightWeiboCell.TOPIC_AREA_HEIGHT + contentSize.size.height + AutoHeightWeiboCell.IMAGE_AREA_HEIGHT + AutoHeightWeiboCell.BOTTOM_PADDING * 2;//返回计算后的最终高度
    }
    
    //MARK: 根据外部传入数据开始布局
    override func showSubviews() {
        self.backgroundColor = UIColor.white
        
        let weiboModel:WeiboModel? = getData() as? WeiboModel//获取Model
        if weiboModel == nil {
            return
        }
        
        self.iconView.sd_setImage(with: URL.init(string: weiboModel!.iconName), completed: nil)
        self.iconView.centerX = AutoHeightWeiboCell.LEFT_PADDING / 2.0
        self.iconView.centerY = AutoHeightWeiboCell.TOPIC_AREA_HEIGHT / 2.0
        
        self.nameLabel.text = weiboModel?.name
        self.nameLabel.sizeToFit()
        self.titleLabel.text = weiboModel?.title
        self.titleLabel.sizeToFit()
        
        let gap:CGFloat = 5
        let baseY:CGFloat = (AutoHeightWeiboCell.TOPIC_AREA_HEIGHT - self.nameLabel.height - gap - self.titleLabel.height) / 2.0
        self.nameLabel.y = baseY
        self.titleLabel.y = self.nameLabel.maxY + gap
        
        self.nameLabel.x = AutoHeightWeiboCell.LEFT_PADDING
        self.titleLabel.x = self.nameLabel.x
        
        self.contentLabel.text = weiboModel?.content;
        self.contentLabel.size = self.contentLabel.sizeThatFits(CGSize.init(width: self.contentView.width - AutoHeightWeiboCell.LEFT_PADDING - AutoHeightWeiboCell.RIGHT_PADDING, height: CGFloat(Float.greatestFiniteMagnitude)))
        self.contentLabel.x = AutoHeightWeiboCell.LEFT_PADDING
        self.contentLabel.y = AutoHeightWeiboCell.TOPIC_AREA_HEIGHT
        
        
        self.photoView.sd_setImage(with: URL.init(string: weiboModel!.imageUrl), completed: nil)
        self.photoView.frame = CGRect.init(x: AutoHeightWeiboCell.LEFT_PADDING, y: self.contentLabel.maxY + AutoHeightWeiboCell.BOTTOM_PADDING, width: self.contentLabel.width, height: AutoHeightWeiboCell.IMAGE_AREA_HEIGHT)
        
        self.bottomLine.x = 0
        self.bottomLine.maxY = self.contentView.height
        self.bottomLine.width = self.contentView.width
    }
    
    //MARK: 懒加载添加视图
    private lazy var iconView:UIImageView = { () -> UIImageView in
        let _iconView:UIImageView = UIImageView()
        _iconView.contentMode = UIView.ContentMode.scaleAspectFill
        let iconWidth:CGFloat = AutoHeightWeiboCell.LEFT_PADDING - 20
        _iconView.size = CGSize.init(width: iconWidth, height: iconWidth)
        _iconView.layer.cornerRadius = iconWidth / 2.0
        _iconView.layer.masksToBounds = true
        self.contentView.addSubview(_iconView)
        return _iconView
    }()
        
    private lazy var photoView:UIImageView = { () -> UIImageView in
        let _photoView:UIImageView = UIImageView()
        _photoView.contentMode = UIView.ContentMode.scaleAspectFill
        _photoView.layer.masksToBounds = true
        self.contentView.addSubview(_photoView)
        return _photoView
    }()
    
    private lazy var titleLabel:UILabel = { () -> UILabel in
        let _titleLabel:UILabel = UICreationUtils.createLabel(TVStyle.sizeTextSecondary, color: TVStyle.colorTextSecondary)
        self.contentView.addSubview(_titleLabel)
        return _titleLabel
    }()
        
    private lazy var nameLabel:UILabel = { () -> UILabel in
        let _nameLabel:UILabel = UICreationUtils.createLabel(TVStyle.sizeTextPrimary, color: TVStyle.colorTextPrimary)
        self.contentView.addSubview(_nameLabel)
        return _nameLabel
    }()
    
    private lazy var contentLabel:UILabel = { () -> UILabel in
        let _contentLabel:UILabel = UICreationUtils.createLabel(TVStyle.sizeTextSecondary, color: TVStyle.colorTextPrimary)
        _contentLabel.lineBreakMode = NSLineBreakMode.byWordWrapping
        _contentLabel.numberOfLines = 0
        self.contentView.addSubview(_contentLabel)
        return _contentLabel
    }()
        
    private lazy var bottomLine:UIView = { () -> UIView in
        let _bottomLine:UIView = UIView()
        _bottomLine.backgroundColor = TVStyle.colorLine
        _bottomLine.height = TVStyle.lineWidth
        self.contentView.addSubview(_bottomLine)
        return _bottomLine
    }()

}
